### ISSUE TITLE
ShardedJedisPool check bug

### ISSUE THREAD

**Issue by Paul1Q84** [xetorthio/jedis#1277]
I found some proplem in ShardedJedisPool when using testWhileIdle,Some Jedis instance in ShardedJedis are never used and connected,but when test it will call validateObject and when idle time more than MinEvictableIdleTimeMillis it will destroyObject,this two place can make some never used Jedis create socket and connected it.some suggest below. is it right?

```
private boolean hasUsed(Jedis jedis){
return jedis.getClient().getSocket() != null;
}
@Override
public void destroyObject(PooledObject pooledShardedJedis) throws Exception {
final ShardedJedis shardedJedis = pooledShardedJedis.getObject();
for (Jedis jedis : shardedJedis.getAllShards()) {
try {
try {
//judge before
if(hasUsed(jedis)) jedis.quit();
} catch (Exception e) {

      }
      jedis.disconnect();
    } catch (Exception e) {

    }
  }
}    
@Override
public boolean validateObject(PooledObject<ShardedJedis> pooledShardedJedis) {
  try {
    ShardedJedis jedis = pooledShardedJedis.getObject();
    for (Jedis shard : jedis.getAllShards()) {
        //judge before
        if (hasUsed(shard) && !shard.ping().equals("PONG")) {
        return false;
      }
    }
    return true;
  } catch (Exception ex) {
    return false;
  }
}
```

#### Jedis version:
2.7.2
#### Redis version:
2.8
#### Java version:
1.7

**Comment by marcosnils** [comment]
**tl; dr**

This is a normal situation that appears in any pool / configuration environment, I don't like adding a `used` logic as it brings some other scenarios that are not considered in the standard usage. 

Recommendations: 

- Tune the pool parameters to suit your needs (in this case increase the `MinEvictableIdleTimeMillis`)
- Use a better hash key / alg so your shards are accessed evenly


@Paul1Q84 This approach would totally break the purpose of `MinEvictableIdleTimeMillis` as it won't destroy the object if it has never been used before. 


In any case in a "sharded" environment it's usually better if all the keys are accessed evenly. If you have some specific keys / nodes that are rarely accessed I guess you'd have to review your hashing key / alg as it might lead to this scenarios you're getting into.

**Comment by Paul1Q84** [comment]
> @Paul1Q84 This approach would totally break the purpose of MinEvictableIdleTimeMillis as it won't destroy the object if it has never been used before.

 I just prevent a never used and connected Jedis to call quit function,because the quit function will create new socket and connected in the detroy function. You will detroy it so why you want to create and connected it?

Sorry my english is so so.

**Comment by marcosnils** [comment]
@Paul1Q84 it is rare that an  instance is never used before the MinEvictableIdleTimeMillis hits. If that is the case I'd recommend either increasing the MinEvictableIdleTimeMillis time or re-evaluating your sharding keys
